File: Book/Java/src/BruteForceSearch.java
public class BruteForceSearch {

    public static void main(String[] args) {

        int[] xs = {0, 1, 2, 3, 4, 5, 6, 7, 8};
        int[] ys = {1, 0, 3, 2, 5, 4, 7, 6, 8};
        int[] zs = {8, 7, 6, 5, 4, 3, 2, 1, 0};

        int xsNInvs = bruteForceSearch(xs);
        int ysNInvs = bruteForceSearch(ys);
        int zsNInvs = bruteForceSearch(zs);

        System.out.println("Xs: ");
        System.out.println(xsNInvs);
        System.out.println("Ys: ");
        System.out.println(ysNInvs);
        System.out.println("Zs: ");
        System.out.println(zsNInvs);

    }

    /**
     * Brute-Force Search for Counting Inversions
     *
     * @param xs
     * Input: array A of n distinct integers.
     *
     * @return numInv
     * Output: the number of inversions of A.
     */
    public static int bruteForceSearch(int[] xs) {

        int numInv = 0;
        int n = xs.length;

        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (xs[i] > xs [j]) {
                    numInv++;
                }
            }
        }

        return numInv;
    }
}

File: Haskell/ListsAndArrays/CountInversions/BruteForceSearch/src/BruteForceSearch.hs
---------------------------------------------------------
-- Module: BruteForceSearch
-- Counts the number of inversions in a list of items.
-- An inversion is an element x > y s.t. appears before 
-- y in the list (to the "left" of y).
---------------------------------------------------------

module BruteForceSearch (bruteForceSearch) where

-- Input: array A of n distinct integers.
-- Output: the number of inversions of A.
bruteForceSearch :: (Ord a) => [a] -> Integer
bruteForceSearch xs =

    outer xs 0

    -- Go through every element, x, of the list.
    where outer [] rsf = rsf
          outer (x:xs) rsf =
            outer xs (rsf + (inner xs 0))

            -- Check for inversions:
            --  Elements that are smaller than x but that appear
            --   later in xs.
            where inner [] rsf = rsf
                  inner (x':xs) rsf =
                    inner xs (if x > x' then (succ rsf)
                                        else rsf)